// @heroui/react-utils/dist/index.mjs
"use client";
import {
  createContext
} from "./chunk-3XT5V4LF.mjs";
import {
  getCSSStyleVal,
  getRealShape
} from "./chunk-VKMEMDAU.mjs";
import {
  areRectsIntersecting,
  canUseDOM,
  createDOMRef,
  createFocusableRef,
  detectBrowser,
  detectDeviceType,
  detectOS,
  detectTouch,
  getUserAgentBrowser,
  getUserAgentOS,
  isBrowser,
  useDOMRef,
  useFocusableRef,
  useSyncRef
} from "./chunk-OEE6MISH.mjs";
import {
  assignRef,
  mergeRefs
} from "./chunk-OTWYT2HS.mjs";
import {
  useIsHydrated
} from "./chunk-6UBKM7F3.mjs";

// src/index.ts
import {
  DOMPropNames,
  DOMEventNames,
  getValidChildren,
  pickChildren,
  renderFn,
  filterDOMProps
} from "@heroui/react-rsc-utils";
export {
  DOMEventNames,
  DOMPropNames,
  areRectsIntersecting,
  assignRef,
  canUseDOM,
  createContext,
  createDOMRef,
  createFocusableRef,
  detectBrowser,
  detectDeviceType,
  detectOS,
  detectTouch,
  filterDOMProps,
  getCSSStyleVal,
  getRealShape,
  getUserAgentBrowser,
  getUserAgentOS,
  getValidChildren,
  isBrowser,
  mergeRefs,
  pickChildren,
  renderFn,
  useDOMRef,
  useFocusableRef,
  useIsHydrated,
  useSyncRef
};
