// @heroui/toast/dist/chunk-XOTBZDRP.mjs
"use client";
import {
  ToastRegion
} from "./chunk-RJHHOZ5N.mjs";

// src/toast-provider.tsx
import { ToastQueue, useToastQueue } from "@react-stately/toast";
import { useProviderContext } from "@heroui/system";
import { jsx } from "react/jsx-runtime";
var globalToastQueue = null;
var getToastQueue = () => {
  if (!globalToastQueue) {
    globalToastQueue = new ToastQueue({
      maxVisibleToasts: Infinity,
      hasExitAnimation: true
    });
  }
  return globalToastQueue;
};
var ToastProvider = ({
  placement = "bottom-right",
  disableAnimation: disableAnimationProp = false,
  maxVisibleToasts = 3,
  toastOffset = 0,
  toastProps = {}
}) => {
  var _a;
  const toastQueue = useToastQueue(getToastQueue());
  const globalContext = useProviderContext();
  const disableAnimation = (_a = disableAnimationProp != null ? disableAnimationProp : globalContext == null ? void 0 : globalContext.disableAnimation) != null ? _a : false;
  if (toastQueue.visibleToasts.length == 0) {
    return null;
  }
  return /* @__PURE__ */ jsx(
    ToastRegion,
    {
      disableAnimation,
      maxVisibleToasts,
      placement,
      toastOffset,
      toastProps,
      toastQueue
    }
  );
};
var addToast = ({ ...props }) => {
  if (!globalToastQueue) {
    return;
  }
  const options = {
    priority: props == null ? void 0 : props.priority
  };
  globalToastQueue.add(props, options);
};
var closeAll = () => {
  if (!globalToastQueue) {
    return;
  }
  const keys = globalToastQueue.visibleToasts.map((toast) => toast.key);
  keys.map((key) => {
    globalToastQueue == null ? void 0 : globalToastQueue.close(key);
  });
};

export {
  getToastQueue,
  ToastProvider,
  addToast,
  closeAll
};
